File: 1930-UniqueLength-3PalindromicSubsequences/1930-UniqueLength-3PalindromicSubsequences.py
# Last updated: 11/5/2025, 6:17:46 pm
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        n = len(s)

        freq = [0] * 26

        arr = [ord(char) - 97 for char in s]
        ret = 0

        left = [0] * 26
        right = [0] * 26
        
        for i in range(1, n):
            right[arr[i]] += 1
        
        left[arr[0]] += 1

        visited = [[1] * 26 for _ in range(26)]
        
        for i in range(1, n):
            a = arr[i]
            right[a] -= 1

            for num in range(26):
                if visited[num][a] and left[num] and right[num]:
                    ret += 1
                    visited[num][a] = 0
            
            left[a] += 1
        
        return ret
            

File: 1930-UniqueLength-3PalindromicSubsequences/1930-UniqueLength-3PalindromicSubsequences.cpp
// Last updated: 21/11/2025, 5:16:41 pm
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        int n = s.length();
        int ans = 0;
        int lowercase = 'a';
        for (int a=lowercase; a<lowercase+26; a++) {
            char c = static_cast<char>(a);
            int left = -1;
            int right = -1;
            for (int i=0; i<n; i++) {
                if (s[i] == c) {
                    if(left == -1) left = i;
                    else right = i;
                }
            }
            unordered_set<char> uniq;
            for (int i=left+1; i<right; i++) {
                uniq.insert(s[i]);
            }
            ans += uniq.size();
        }

        return ans;
    }
};